pkg/kubelet/cm/topologymanager/policy_single_numa_node.go

```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
 	}
 }
 
+// Get the narrowest hint aligned to one NUMA node
+// assumption: allResourcesHints are not empty, all hints have NUMAAffinity != nil,
+// all hints have exactly one NUMA node set in NUMANodeAffinity
+func (p *singleNumaNodePolicy) getHintMatch(allResourcesHints [][]TopologyHint) (bool, TopologyHint) {
+	// Sort all resources hints
+	for _, resource := range allResourcesHints {
+		sort.Slice(resource, func(i, j int) bool {
+			if resource[i].NUMANodeAffinity.GetBits()[0] < resource[j].NUMANodeAffinity.GetBits()[0] {
+				return true
+			}
+			return false
+		})
+
+	}
+	// find a match by searching a hint of one resource in the rest
+	var match TopologyHint
+	var foundMatch bool
+
+	if len(allResourcesHints) == 1 {
+		match = allResourcesHints[0][0]
+		match.Preferred = true
+		return true, match
+	}
+	for _, candidate := range allResourcesHints[0] {
+		foundMatch = true
+		for _, hints := range allResourcesHints[1:] {
+			res := sort.Search(len(hints), func(i int) bool {
+				return hints[i].NUMANodeAffinity.GetBits()[0] >= candidate.NUMANodeAffinity.GetBits()[0]
+			})
+			if res >= len(hints) ||
+				!hints[res].NUMANodeAffinity.IsEqual(candidate.NUMANodeAffinity) {
+				// hint not found, move to next hint from allResourcesHints[0]
+				foundMatch = false
+				break
+			}
+		}
+		if foundMatch {
+			match = candidate
+			match.Preferred = true
+			break
+		}
+	}
+	return foundMatch, match
+}
+
+// Return hints that have valid bitmasks with exactly one bit set
+func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [][]TopologyHint {
+	var filteredResourcesHints [][]TopologyHint
+	if len(allResourcesHints) > 0 {
+		for _, oneResourceHints := range allResourcesHints {
+			var filtered []TopologyHint
+			if len(oneResourceHints) > 0 {
+				for _, hint := range oneResourceHints {
+					if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 {
+						filtered = append(filtered, hint)
+					}
+				}
+			}
+			filteredResourcesHints = append(filteredResourcesHints, filtered)
+		}
+	}
+	return filteredResourcesHints
+}
+
```

Comment:
The standard pattern we tend to follow in kubernetes is to return early based on a condition, rather than having the code structured as you do here. This would transform this function
```
func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [][]TopologyHint {
        if len(allResourcesHints) == 0 {
                return nil
        }

        var filteredResourcesHints [][]TopologyHint
        for _, oneResourceHints := range allResourcesHints {
                if len(oneResourceHints) == 0 {
                        continue
                }

                var filtered []TopologyHint
                for _, hint := range oneResourceHints {
                        if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 {
                                filtered = append(filtered, hint)
                        }
                }

                filteredResourcesHints = append(filteredResourcesHints, filtered)
        }

        return filteredResourcesHints
}
```

Reply:
Just testing this out now, it will need a couple of minor tweaks to produce the desired output:
``````

func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [][]TopologyHint {
        var filteredResourcesHints [][]TopologyHint
        for _, oneResourceHints := range allResourcesHints {
                var filtered []TopologyHint
                if len(oneResourceHints) == 0 {
                        filteredResourcesHints = append(filteredResourcesHints, filtered)
                        continue
                }

                for _, hint := range oneResourceHints {
                        if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 && hint.Preferred == true {
                                filtered = append(filtered, hint)
                        }
                }

                filteredResourcesHints = append(filteredResourcesHints, filtered)
        }
        return filteredResourcesHints
}

Reply:
I guess it depends on what the desired output is.

My original suggestion would have output exactly what the previous iteration of the code would output. Your new version now *always* does a `filteredResourcesHints = append(filteredResourcesHints, filtered)` in all cases (even if `filtered` is `nil`), which wouldn't have happened before.

With this new version (assuming these are the desired semantics), the whole check for `if len(oneResourceHints) == 0` can actually be removed (the `filteredResourcesHints = append(filteredResourcesHints, filtered)` at the bottom should suffice for both cases).

Reply:
Maybe the true desired semantics arethe following?

```
func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [ []TopologyHint {
        var filteredResourcesHints [][]TopologyHint
        for _, oneResourceHints := range allResourcesHints {
                var filtered []TopologyHint
                for _, hint := range oneResourceHints {
                        if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 && hint.Preferred == true {
                                filtered = append(filtered, hint)
                        }
                }
                if len(filtered) > 0 {
                        filteredResourcesHints = append(filteredResourcesHints, filtered)
                }
        }
        return filteredResourcesHints
}
```

Reply:
> I guess it depends on what the desired output is.
> 
> My original suggestion would have output exactly what the previous iteration of the code would output. Your new version now _always_ does a `filteredResourcesHints = append(filteredResourcesHints, filtered)` in all cases (even if `filtered` is `nil`), which wouldn't have happened before.
> 
> With this new version (assuming these are the desired semantics), the whole check for `if len(oneResourceHints) == 0` can actually be removed (the `filteredResourcesHints = append(filteredResourcesHints, filtered)` at the bottom should suffice for both cases).

I might be missing something here, but the original function does in fact always do a `filteredResourcesHints = append(filteredResourcesHints, filtered)` in all cases (even if `filtered` is `nil`)? 

The idea being that there is then a check for any `nil` hint after `filterHints` is called: https://github.com/kubernetes/kubernetes/blob/078f754a33bb0fcf2a271c2d95ca16a9d9b81c0a/pkg/kubelet/cm/topologymanager/policy_single_numa_node.go#L174

The version I added achieves the same output as the original.

Reply:
There's now too many versions of this to keep track of :) Not sure which one you mean when you say original anymore.

If we do want `nil` added, then it's fine to get rid of the check for `if len(filtered) > 0`, but I still think it's better to omit the check for the following because it is unnecessary:
```
                if len(oneResourceHints) == 0 {
                        filteredResourcesHints = append(filteredResourcesHints, filtered)
                        continue
                }
```
If `len(oneResourceHints) == 0` then the loop just below it won't be entered and we will fall down to the `filteredResourcesHints = append(filteredResourcesHints, filtered)` and perform the same operation as whats inside this conditional.

Reply:
Yes, I think we're on the same page now! 😄 I'll tidy this function up.